src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1499 @@ void Init()
+    public string? Value { get; }
+
+    public string M()
+    {
+        if (IsOk is false)
+        {
+            return Value; // 1
+        }
+        else
+        {
+            return Value;
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (15,20): warning CS8603: Possible null reference return.
+                //             return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(15, 20)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchStatementOnProperty()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (IsOk)
+        {
+            case true:
+                return Value;
+            default:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnProperty()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk switch
+        {
+            true => Value,
+            _ => throw null!,
+        };
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnMethod()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk() { throw null!; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        return IsOk() switch
+        {
+            true => Value,
+            _ => throw null!,
+        };
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_SwitchExpressionOnProperty_WhenFalseBranch()
```

Comment:
>MemberNotNullWhenTrue_SwitchExpressionOnProperty_WhenFalseBranch [](start = 20, length = 64)

Consider merging into `MemberNotNullWhenTrue_SwitchExpressionOnProperty`.

Reply:
Same problem. The diagnostic is on the return, but I want to show which branch the problem comes from.

---
In reply to: [573116282](https://github.com/dotnet/roslyn/pull/51001#discussion_r573116282) [](ancestors = 573116282)